src/internal.rs

```diff
@@ -410,82 +412,100 @@ impl<I, O1, O2, E, F: Parser<I, O1, E>, G: FnMut(O1) -> O2> Parser<I, O2, E> for
 }
 
 /// Implementation of `Parser::map_res`
-pub struct MapRes<F, G, O> {
+pub struct MapRes<F, G> {
   f: F,
   g: G,
-  phantom: core::marker::PhantomData<O>,
 }
 
-impl<I, O, O2, E, E2, F, G> Parser<I, O2, E> for MapRes<F, G, O>
+impl<I, O2, E2, F, G> Parser<I> for MapRes<F, G>
 where
   I: Clone,
-  E: FromExternalError<I, E2>,
-  F: Parser<I, O, E>,
-  G: Fn(O) -> Result<O2, E2>,
+  <F as Parser<I>>::Error: FromExternalError<I, E2>,
+  F: Parser<I>,
+  G: Fn(<F as Parser<I>>::Output) -> Result<O2, E2>,
 {
-  fn parse(&mut self, input: I) -> IResult<I, O2, E> {
+  type Output = O2;
+  type Error = <F as Parser<I>>::Error;
+  fn parse(&mut self, input: I) -> IResult<I, O2, <F as Parser<I>>::Error> {
     let i = input.clone();
     let (input, o1) = self.f.parse(input)?;
     match (self.g)(o1) {
       Ok(o2) => Ok((input, o2)),
-      Err(e) => Err(Err::Error(E::from_external_error(i, ErrorKind::MapRes, e))),
+      Err(e) => Err(Err::Error(<F as Parser<I>>::Error::from_external_error(
+        i,
+        ErrorKind::MapRes,
+        e,
+      ))),
     }
   }
 }
 
 /// Implementation of `Parser::map_opt`
-pub struct MapOpt<F, G, O> {
+pub struct MapOpt<F, G> {
   f: F,
   g: G,
-  phantom: core::marker::PhantomData<O>,
 }
 
-impl<I, O, O2, E, F, G> Parser<I, O2, E> for MapOpt<F, G, O>
+impl<I, O2, F, G> Parser<I> for MapOpt<F, G>
 where
   I: Clone,
-  E: ParseError<I>,
-  F: Parser<I, O, E>,
-  G: Fn(O) -> Option<O2>,
+  F: Parser<I>,
+  G: Fn(<F as Parser<I>>::Output) -> Option<O2>,
 {
-  fn parse(&mut self, input: I) -> IResult<I, O2, E> {
+  type Output = O2;
+  type Error = <F as Parser<I>>::Error;
+
+  fn parse(&mut self, input: I) -> IResult<I, O2, <F as Parser<I>>::Error> {
     let i = input.clone();
     let (input, o1) = self.f.parse(input)?;
     match (self.g)(o1) {
       Some(o2) => Ok((input, o2)),
-      None => Err(Err::Error(E::from_error_kind(i, ErrorKind::MapOpt))),
+      None => Err(Err::Error(<F as Parser<I>>::Error::from_error_kind(
+        i,
+        ErrorKind::MapOpt,
+      ))),
     }
   }
 }
 
 /// Implementation of `Parser::flat_map`
 #[cfg_attr(nightly, warn(rustdoc::missing_doc_code_examples))]
-pub struct FlatMap<F, G, O1> {
+pub struct FlatMap<F, G> {
   f: F,
   g: G,
-  phantom: core::marker::PhantomData<O1>,
```

Comment:
Nice that this lets you drop the phantom data

Reply:
yes, it simplifies them nicely. I have to check again the methods like `flat_map`, to make sure that they require all the trait bounds, otherwise it would be possible for them to return a value that does not implement `Parser`, which would be confusing